File: lib/manufactured/event_handler.rb
# Manufactured Event Handler Definition
#
# Copyright (C) 2014 Mohammed Morsi <mo@morsi.org>
# Licensed under the AGPLv3+ http://www.gnu.org/licenses/agpl.txt

require 'omega/server/event'
require 'manufactured/events'

module Manufactured

# Subclasses Omega::Server::EventHandler to define a custom
# event handler with more granular event matching
class EventHandler < Omega::Server::EventHandler
  attr_accessor :event_args

  def initialize(args={})
    attr_from_args args, :event_args => []
    super(args)
  end

  def matches?(event)
    super(event) && event.trigger_handler?(self)
  end

  def to_json(*a)
    { 'json_class' => self.class.name,
      'data'       =>
        json_data.merge({:event_args => event_args})
    }.to_json(*a)
  end

end # class EventHandler
end # module Manufactured

File: site/source/javascripts/pages/mixins/event_handler.js
/* Omega Page EventHandler Mixin
 *
 * Copyright (C) 2014 Mohammed Morsi <mo@morsi.org>
 * Licensed under the AGPLv3 http://www.gnu.org/licenses/agpl.txt
 */

Omega.Pages.EventHandler = {
  /// Helper to register handlers for all supported events
  _handle_events : function(){
    var events = Omega.CallbackHandler.all_events();
    for(var e = 0; e < events.length; e++)
      this.callback_handler.track(events[e]);
  }
};
